File: primalTestDotNet/AI/States/LookingForSwordState.cs
using PrimalTestDotNet.Entities;

namespace PrimalTestDotNet.AI.States;

public class LookingForSwordState(AIStateMachine context, Level.Level level) : StateBase(level), IState
{
    public void Update(Hero hero, IEnumerable<IGameObject> gameObjects)
    {
        if (hero.HasSword)
        {
            context.ChangeState(context.LookingForTreasureState);
            return;
        }

        // Look for a sword without monsters on the path
        var allMonsters = gameObjects.Where(o => o is Monster);
        var sword = gameObjects.Where(o => o is Sword).FirstOrDefault();
        if (sword == null)
        {
            context.ChangeState(context.UnwinnableState);
            return;
        }
        // Consider monsters as walls in that case
        var recalculatedDistanceData = CalculateDistanceData(hero.Position, allMonsters);
        var pathToSword = GetPath(hero.Position, sword.Position, recalculatedDistanceData);

        var node = pathToSword.Last();
        var direction = node - hero.Position;
        hero.Move(direction);
    }
}

File: PrimalTestCpp/LookingForSwordState.cpp
#include "LookingForSwordState.h"
#include "Hero.h"
#include "AIStateMachine.h"
#include "Monster.h"
#include "Sword.h"

LookingForSwordState::LookingForSwordState(AIStateMachine& context, Level& level)
	: StateBase(level), _context(context)
{
}

void LookingForSwordState::update(Hero& hero, std::vector<std::shared_ptr<IGameObject>>& gameObjects)
{
	if (hero.hasSword()) {
		_context.changeState(_context.LOOKING_FOR_TREASURE_STATE());
		return;
	}

	// Look for sword without monsters on the path
	auto allMonsters = findAllObjectsOfType<Monster>(gameObjects);
	auto swords = findAllObjectsOfType<Sword>(gameObjects);
	// No sword found, but there are monsters in the way => forfeit
	if (swords.size() == 0) {
		_context.changeState(_context.UNWINNABLE_STATE());
		return;
	}

	// monsters are considered as walls when the hero does not have a sword
	auto& sword = swords.front();
	auto recalculatedDistanceData = calculateDistanceData(hero.getPosition(), allMonsters);
	auto pathToSword = getPath(hero.getPosition(), sword->getPosition(), recalculatedDistanceData);

	// move to sword
	auto& node = pathToSword.back();
	auto direction = node - hero.getPosition();
	hero.move(direction);
}
